PostgreSQL 12 + PostGIS 3 + GEOS 3.8.1 + Python 3.8.6 tests

espinafre opened the pull request:
I've noticed that lately the PostgreSQL tests have failing on me, as I've explained on #39401. One of the problems is a rounding error, which I've fixed by adding round() to 4 digits on the relevant PostgreSQL test cases; the other issue was a little harder.

I've bisected it to  250ce7615293b3970cc6a98ec925e2a22b1608a0, which introduced the value of 8.0 as the default parameter "segments" for the buffer() QGIS function. Since that literal is a floating point value, the PostgreSQL provider quotes it (per #37367), which makes the failing test generate the following SQL snippet:

`ST_Buffer("geom",1,'8')`

The PostGIS [documentation](https://postgis.net/docs/ST_Buffer.html) says that ST_Buffer has two signatures for geometry types:

```
geometry ST_Buffer(geometry g1, float radius_of_buffer, text buffer_style_parameters='');
geometry ST_Buffer(geometry g1, float radius_of_buffer, integer num_seg_quarter_circle);
```

Since the value '8' is passed quoted to the DBMS, PostGIS understands that it is a string, and the first form is called, which generates the error and ultimately causes the test to fail.

With this change, the value 8.0 is now simply 8, which is treated as an integer by the provider, and therefore not quoted, which generates the following SQL snippet:

`ST_Buffer("geom",1,8)`

which is the second form of the ST_Buffer function documented above, and what we really want. Also, it makes more sense that the number of segments in a quarter circle is integer!

Fixes  #39401

espinafre committed:
Number of segments in buffer() function is integer

This makes the parameter that indicates the number of segments in a
quarter circle for the buffer() function an integer instead of a
floating point value. When it is input as a floating point, the
PostgreSQL provider quotes it, making it appear as a text, causing a
parameter error on calling PostGIS's ST_Buffer().

espinafre committed:
round() on float equality comparisons in pgsql tests

We round floating point values in tests that check for equality between
two values in PostgreSQL tests, so that floating point implementation
differences in Python, PostgreSQL, GEOS etc. can be sorted out.

nyalldawson commented:
> I've bisected it to 250ce76, which introduced the value of 8.0 as the default parameter "segments" for the buffer() QGIS function. Since that literal is a floating point value, the PostgreSQL provider quotes it (per #37367), which makes the failing test generate the following SQL snippet:

Ewww -- #37367 is completely wrong in changing this! Instead of changing the expression function we need to partially rollback #37367 and ensure that double values AREN'T quoted in the postgres expression compiler. Are you able to do this?

nyalldawson commented on tests/src/python/test_provider_postgres.py:
```diff
@@ -554,7 +554,7 @@ def testGeneratedFields(self):
 
         assert compareWkt(generated_geometry, expected_geometry), "Geometry mismatch! Expected:\n{}\nGot:\n{}\n".format(expected_geometry, generated_geometry)
-        self.assertEqual(f2['poly_area'], expected_area)
+        self.assertEqual(round(f2['poly_area'], 4), round(expected_area, 4))
```
You can use self.assertAlmostEqual here instead, which will give nicer debug messages when the test fails

espinafre replied:
I'll try this, thanks for the pointer!

espinafre replied:
Fixed by 9537f48

espinafre commented:
> > I've bisected it to [250ce76](https://github.com/qgis/QGIS/commit/250ce7615293b3970cc6a98ec925e2a22b1608a0), which introduced the value of 8.0 as the default parameter "segments" for the buffer() QGIS function. Since that literal is a floating point value, the PostgreSQL provider quotes it (per #37367), which makes the failing test generate the following SQL snippet:
> 
> Ewww -- #37367 is completely wrong in changing this! Instead of changing the expression function we need to partially rollback #37367 and ensure that double values AREN'T quoted in the postgres expression compiler. Are you able to do this?

Yes, I certainly can do that; however, we lose the ability to use fixed point (NUMERIC in PostgreSQL) columns as primary keys, or as part of a composite primary key, as I've shown in issue #37245 . Are you okay with this? Shall I proceed? Thanks!

nyalldawson commented:
> . Are you okay with this? Shall I proceed? Thanks!

Not if we limit the changes from the earlier PR so that they don't apply to the postgres SQL expression compiler. This is only used for conversion of QGIS filter expressions to postgres expressions, so won't affect the PK related filtering code.

espinafre committed:
pgsql12: using assertAlmostEqual in floating point tests.

On suggestion by Nyall Dawson, we use assertAlmostEqual instead of
assertEqual+round() on floating point comparisons in the PostgreSQL
provider tests.

espinafre committed:
pgsql12 tests: Python 3.6 compatibility

espinafre committed:
pgsql12 tests: Floating point comparison precision.

espinafre commented:
Hi, @nyalldawson , I couldn't find a way to leave floating point literals unquoted just for the expression compiler; it seems to me that it is an all-or-nothing proposition. There might be a way, but I didn't find it. As it is, this PR at least allows the tests to pass against PostgreSQL 12 (and now, against PostgreSQL 13) on my local systems.

Can we merge this while we further investigate or devise a form to separate expression compiler's statements from the rest of them?

nyalldawson commented:
@espinafre I've fixed the expression compiler in https://github.com/qgis/QGIS/pull/40713
